org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/IndexUtils.java

```diff
@@ -0,0 +1,313 @@
+import java.util.Objects;
+import java.util.Set;
+import java.util.zip.CRC32;
+
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResourceChangeEvent;
+import org.eclipse.core.resources.IResourceChangeListener;
+import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.internal.core.ClasspathEntry;
+import org.eclipse.jdt.internal.core.JavaModel;
+import org.eclipse.jdt.internal.core.JavaModelManager;
+import org.eclipse.jdt.internal.core.JavaProject;
+import org.eclipse.jdt.internal.core.index.FileIndexLocation;
+import org.eclipse.jdt.internal.core.index.Index;
+import org.eclipse.jdt.internal.core.index.IndexLocation;
+import org.eclipse.jdt.internal.core.search.indexing.IndexManager;
+import org.eclipse.jdt.internal.core.search.indexing.ReadWriteMonitor;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.JobHelpers;
+import org.eclipse.jdt.ls.core.internal.ProjectUtils;
+
+public class IndexUtils {
+	private static boolean resourceChangeRegistered = false;
+	private static Map<IPath, Long> externalTimeStamps = null;
+
+	public static void copyIndexesToSharedLocation() {
+		// common index location for all workspaces
+		final String SHARED_INDEX_LOCATION = System.getProperty("jdt.core.sharedIndexLocation");
+		if (JavaModelManager.getIndexManager() == null || StringUtils.isBlank(SHARED_INDEX_LOCATION)) {
+			return;
+		}
+
+		JobHelpers.waitUntilIndexesReady();
+		getExternalLibTimeStamps(); // init load of externalLibTimeStamps
+		registerResourceChangeListener();
+		copyIndexesToSharedLocation(ProjectUtils.getJavaProjects());
+	}
+
+	private static synchronized void registerResourceChangeListener() {
+		if (resourceChangeRegistered) {
+			return;
+		}
+		resourceChangeRegistered = true;
+		ResourcesPlugin.getWorkspace().addResourceChangeListener(new IResourceChangeListener() {
+			@Override
+			public void resourceChanged(IResourceChangeEvent event) {
+				IJavaProject [] projects = null;
+				Object obj = event.getSource();
+				if (obj instanceof IProject project) {
+					if (ProjectUtils.isJavaProject(project)) {
+						projects = new IJavaProject[]{ JavaCore.create(project) };
+					}
+				} else if (obj instanceof IWorkspace) {
+					projects = ProjectUtils.getJavaProjects();
+				}
+
+				if (projects != null && projects.length > 0) {
+					JavaModelManager.getIndexManager().waitForIndex(true, null);
+					copyIndexesToSharedLocation(projects);
+				}
+			}
+		}, IResourceChangeEvent.PRE_REFRESH);
+	}
+
+	private static void copyIndexesToSharedLocation(IJavaProject[] javaProjects) {
```

Comment:
I'm thinking that if it is necessary to schedule this method into a Job, since it's dealing with IO and probably takes a couple of seconds to finish.

Reply:
What benefits do you expect from wrapping it as a Job? Lock the workspace? Or show some progress notification through the Job?

Reply:
My initial thought is to wrap it to a job with a very low priority - not sure how much benefit will get.